kimchi/src/folding/mod.rs

```diff
@@ -0,0 +1,319 @@
+mod instance_witness;
+#[cfg(test)]
+mod test;
+
+pub(crate) trait FoldingEnv<F> {
+    fn witness(&self, i: usize, side: bool) -> &Vec<F>;
+    fn witness_ext(&self, i: usize, side: bool) -> &Vec<F>;
+    fn index(&self, i: &GateType) -> &Vec<F>;
+    fn coefficient(&self, i: usize) -> &Vec<F>;
+    fn shift(&self) -> &Vec<F>;
+    fn lagrange_basis(&self, i: &i32) -> Vec<F>;
+    ///a vec of just zeros of the same length as other columns
+    fn zero_vec(&self) -> Vec<F>;
+}
+
+pub(crate) enum EvalLeaf<'a, F: Field> {
+    Const(F),
+    Col(&'a Vec<F>),
+    Result(Vec<F>),
+}
+
+impl<'a, F: Field> EvalLeaf<'a, F> {
+    fn map<M: Fn(&F) -> F, I: Fn(&mut F)>(self, map: M, in_place: I) -> Self {
+        use EvalLeaf::*;
+        match self {
+            Const(c) => Const(map(&c)),
+            Col(col) => {
+                let res = col.iter().map(map).collect();
+                Result(res)
+            }
+            Result(mut col) => {
+                for cell in col.iter_mut() {
+                    in_place(cell);
+                }
+                Result(col)
+            }
+        }
+    }
+    fn bin_op<M: Fn(&F, &F) -> F, I: Fn(&mut F, &F)>(
+        a: Self,
+        b: Self,
+        map: M,
+        in_place: I,
+    ) -> Self {
+        use EvalLeaf::*;
+        match (a, b) {
+            (Const(a), Const(b)) => Const(map(&a, &b)),
+            (Const(a), Col(b)) | (Col(b), Const(a)) => {
+                let res = b.iter().map(|f| map(f, &a)).collect();
+                Result(res)
+            }
+            (Col(a), Col(b)) => {
+                let res = a.iter().zip(b.iter()).map(|(a, b)| map(a, b)).collect();
+                Result(res)
+            }
+            (Result(mut a), Const(b)) | (Const(b), Result(mut a)) => {
+                for a in a.iter_mut() {
+                    in_place(a, &b);
+                }
+                Result(a)
+            }
+            (Result(mut a), Col(b)) | (Col(b), Result(mut a)) => {
+                for (a, b) in a.iter_mut().zip(b.iter()) {
+                    in_place(a, b);
+                }
+                Result(a)
+            }
+            (Result(mut a), Result(b)) => {
+                for (a, b) in a.iter_mut().zip(b.iter()) {
+                    in_place(a, b);
+                }
+                Result(a)
+            }
+        }
+    }
+}
+struct Env<'a, G, I, W>
+where
+    G: CommitmentCurve,
+    I: InstanceTrait<G>,
+    W: WitnessTrait<G>,
+{
+    a: (RelaxedInstance<G, I>, RelaxedWitness<G, W>),
+    b: (RelaxedInstance<G, I>, RelaxedWitness<G, W>),
+    index: &'a HashMap<GateType, Vec<G::ScalarField>>,
+    coefficients: &'a [Vec<G::ScalarField>],
+    shift: &'a Vec<G::ScalarField>,
+    len: usize,
+}
+impl<'a, G, I, W> FoldingEnv<G::ScalarField> for Env<'a, G, I, W>
+where
+    G: CommitmentCurve,
+    I: InstanceTrait<G>,
+    W: WitnessTrait<G>,
+{
+    fn witness(&self, i: usize, side: bool) -> &Vec<G::ScalarField> {
+        match side {
```

Comment:
Yeah, agreeing with Danny's question, `side` could be a two-valued enum (Left/Right). This is unclear.

Reply:
will change